File: _DEMO/DemoPlayer.cs
using UnityEngine;
using UnityEngine.InputSystem;
using System;

namespace DEMO
{
    /// <summary>
    /// Replay component for playing back demo files.
    /// Reconstructs gameplay from recorded keyframes and events.
    /// Uses Unity's new Input System (Unity 6000 compatible).
    /// </summary>
    public class DemoPlayer : MonoBehaviour
    {
        [Header("Playback Settings")]
        [SerializeField] private float playbackSpeed = 1f;
        [SerializeField] private bool loopPlayback = false;

        [Header("Camera")]
        [SerializeField] private bool useFreeCamera = true;
        [SerializeField] private float freeCameraSpeed = 10f;

        private DemoFile _currentDemo;
        private bool _isPlaying;
        private bool _isPaused;
        private float _playbackTime;
        private int _currentFrameIndex;

        // Replay objects
        private GameObject _replayPlayer;
        private GameObject _replayCamera;

        // Input System for playback controls
        private InputAction _pauseAction;
        private InputAction _stopAction;
        private InputAction _seekBackAction;
        private InputAction _seekForwardAction;
        private InputAction _speedUpAction;
        private InputAction _speedDownAction;

        // Events
        public event Action OnPlaybackStarted;
        public event Action OnPlaybackStopped;
        public event Action<float> OnPlaybackTimeChanged;

        public bool IsPlaying => _isPlaying;
        public bool IsPaused => _isPaused;
        public float PlaybackTime => _playbackTime;
        public float Duration => _currentDemo?.header.duration ?? 0f;
        public float Progress => Duration > 0 ? _playbackTime / Duration : 0f;

        private void Start()
        {
            SetupInputActions();
        }

        private void OnDestroy()
        {
            CleanupInputActions();
        }

        private void SetupInputActions()
        {
            // Create input actions for playback controls using InputActionMap
            var actionMap = new InputActionMap("DemoPlayback");

            _pauseAction = actionMap.AddAction("Pause", InputActionType.Button);
            _pauseAction.AddBinding("<Keyboard>/space");

            _stopAction = actionMap.AddAction("Stop", InputActionType.Button);
            _stopAction.AddBinding("<Keyboard>/escape");

            _seekBackAction = actionMap.AddAction("SeekBack", InputActionType.Button);
            _seekBackAction.AddBinding("<Keyboard>/leftArrow");

            _seekForwardAction = actionMap.AddAction("SeekForward", InputActionType.Button);
            _seekForwardAction.AddBinding("<Keyboard>/rightArrow");

            _speedUpAction = actionMap.AddAction("SpeedUp", InputActionType.Button);
            _speedUpAction.AddBinding("<Keyboard>/plus");
            _speedUpAction.AddBinding("<Keyboard>/equals");

            _speedDownAction = actionMap.AddAction("SpeedDown", InputActionType.Button);
            _speedDownAction.AddBinding("<Keyboard>/minus");

            // Enable the action map
            actionMap.Enable();
        }

        private void CleanupInputActions()
        {
            _pauseAction?.Disable();
            _stopAction?.Disable();
            _seekBackAction?.Disable();
            _seekForwardAction?.Disable();
            _speedUpAction?.Disable();
            _speedDownAction?.Disable();
        }

        private void Update()
        {
            if (!_isPlaying || _isPaused)
                return;

            UpdatePlayback();
            HandlePlaybackControls();
        }

        /// <summary>
        /// Load a demo file for playback.
        /// </summary>
        public void LoadDemo(DemoFile demo)
        {
            if (demo == null)
            {
                Debug.LogError("[DemoPlayer] Cannot load null demo");
                return;
            }

            _currentDemo = demo;
            _playbackTime = 0f;
            _currentFrameIndex = 0;

            Debug.Log($"[DemoPlayer] Loaded demo: {demo.header.gameVersion}, Duration: {demo.header.duration:F2}s, Frames: {demo.header.totalFrames}");
        }

        /// <summary>
        /// Start playback of the loaded demo.
        /// </summary>
        public void Play()
        {
            if (_currentDemo == null)
            {
                Debug.LogError("[DemoPlayer] No demo loaded");
                return;
            }

            if (_isPlaying && !_isPaused)
                return;

            if (!_isPlaying)
            {
                SetupReplayScene();
                OnPlaybackStarted?.Invoke();
            }

            _isPlaying = true;
            _isPaused = false;
            Debug.Log("[DemoPlayer] Playback started");
        }

        /// <summary>
        /// Pause playback.
        /// </summary>
        public void Pause()
        {
            if (!_isPlaying)
                return;

            _isPaused = true;
            Debug.Log("[DemoPlayer] Playback paused");
        }

        /// <summary>
        /// Resume playback.
        /// </summary>
        public void Resume()
        {
            if (!_isPlaying || !_isPaused)
                return;

            _isPaused = false;
            Debug.Log("[DemoPlayer] Playback resumed");
        }

        /// <summary>
        /// Stop playback and clean up.
        /// </summary>
        public void Stop()
        {
            if (!_isPlaying)
                return;

            _isPlaying = false;
            _isPaused = false;
            _playbackTime = 0f;
            _currentFrameIndex = 0;

            CleanupReplayScene();
            OnPlaybackStopped?.Invoke();

            Debug.Log("[DemoPlayer] Playback stopped");
        }

        /// <summary>
        /// Seek to a specific time in the demo.
        /// </summary>
        public void Seek(float time)
        {
            if (_currentDemo == null)
                return;

            time = Mathf.Clamp(time, 0f, Duration);
            _playbackTime = time;

            // Find the frame at this time
            _currentFrameIndex = FindFrameAtTime(time);

            OnPlaybackTimeChanged?.Invoke(time);
        }

        /// <summary>
        /// Set playback speed.
        /// </summary>
        public void SetPlaybackSpeed(float speed)
        {
            playbackSpeed = Mathf.Max(0.1f, speed);
        }

        /// <summary>
        /// Update playback state each frame.
        /// </summary>
        private void UpdatePlayback()
        {
            if (_currentDemo == null || _currentDemo.frames.Length == 0)
                return;

            // Advance playback time
            _playbackTime += Time.deltaTime * playbackSpeed;

            // Check if playback is complete
            if (_playbackTime >= Duration)
            {
                if (loopPlayback)
                {
                    _playbackTime = 0f;
                    _currentFrameIndex = 0;
                }
                else
                {
                    Stop();
                    return;
                }
            }

            // Find and apply current frame
            _currentFrameIndex = FindFrameAtTime(_playbackTime);

            if (_currentFrameIndex >= 0 && _currentFrameIndex < _currentDemo.frames.Length)
            {
                ApplyFrame(_currentDemo.frames[_currentFrameIndex]);
            }

            OnPlaybackTimeChanged?.Invoke(_playbackTime);
        }

        /// <summary>
        /// Find the frame index at a specific time.
        /// </summary>
        private int FindFrameAtTime(float time)
        {
            if (_currentDemo == null || _currentDemo.frames.Length == 0)
                return 0;

            // Binary search for efficiency
            int left = 0;
            int right = _currentDemo.frames.Length - 1;

            while (left <= right)
            {
                int mid = (left + right) / 2;
                float frameTime = _currentDemo.frames[mid].timestamp;

                if (frameTime < time)
                    left = mid + 1;
                else if (frameTime > time)
                    right = mid - 1;
                else
                    return mid;
            }

            return Mathf.Clamp(left, 0, _currentDemo.frames.Length - 1);
        }

        /// <summary>
        /// Apply frame data to replay objects.
        /// </summary>
        private void ApplyFrame(DemoFrame frame)
        {
            if (_replayPlayer == null)
                return;

            // Apply player state
            _replayPlayer.transform.position = frame.position;
            _replayPlayer.transform.rotation = frame.rotation;

            Rigidbody rb = _replayPlayer.GetComponent<Rigidbody>();
            if (rb != null)
            {
                rb.linearVelocity = frame.velocity;
            }

            // TODO: Apply inputs to player controller if using input-based replay
            // TODO: Handle projectile and target events
        }

        /// <summary>
        /// Setup the replay scene with ghost objects.
        /// </summary>
        private void SetupReplayScene()
        {
            // Create replay player ghost
            _replayPlayer = new GameObject("ReplayPlayer");
            _replayPlayer.transform.position = _currentDemo.playerStartPosition;
            _replayPlayer.transform.rotation = _currentDemo.playerStartRotation;

            // Add visual representation
            // TODO: Add player mesh or capsule renderer

            Rigidbody rb = _replayPlayer.AddComponent<Rigidbody>();
            rb.isKinematic = true;  // We manually control position
            rb.linearVelocity = _currentDemo.playerStartVelocity;

            // Create free camera if enabled
            if (useFreeCamera)
            {
                _replayCamera = new GameObject("ReplayCamera");
                Camera cam = _replayCamera.AddComponent<Camera>();
                cam.depth = 100;  // Render on top
                _replayCamera.transform.position = _currentDemo.playerStartPosition + Vector3.back * 10f + Vector3.up * 3f;
                _replayCamera.transform.LookAt(_currentDemo.playerStartPosition);

                // TODO: Add free camera controls
            }

            // TODO: Create projectile ghosts
            // TODO: Create target/ball ghosts
        }

        /// <summary>
        /// Clean up replay scene objects.
        /// </summary>
        private void CleanupReplayScene()
        {
            if (_replayPlayer != null)
            {
                Destroy(_replayPlayer);
                _replayPlayer = null;
            }

            if (_replayCamera != null)
            {
                Destroy(_replayCamera);
                _replayCamera = null;
            }
        }

        /// <summary>
        /// Handle keyboard playback controls using new Input System.
        /// </summary>
        private void HandlePlaybackControls()
        {
            // Space: Pause/Resume
            if (_pauseAction != null && _pauseAction.WasPressedThisFrame())
            {
                if (_isPaused)
                    Resume();
                else
                    Pause();
            }

            // Escape: Stop
            if (_stopAction != null && _stopAction.WasPressedThisFrame())
            {
                Stop();
            }

            // Arrow keys: Seek
            if (_seekBackAction != null && _seekBackAction.IsPressed())
            {
                Seek(_playbackTime - Time.deltaTime * 2f);
            }
            if (_seekForwardAction != null && _seekForwardAction.IsPressed())
            {
                Seek(_playbackTime + Time.deltaTime * 2f);
            }

            // +/-: Adjust playback speed
            if (_speedUpAction != null && _speedUpAction.WasPressedThisFrame())
            {
                SetPlaybackSpeed(playbackSpeed + 0.25f);
            }
            if (_speedDownAction != null && _speedDownAction.WasPressedThisFrame())
            {
                SetPlaybackSpeed(playbackSpeed - 0.25f);
            }
        }
    }
}

File: client/src/demo/playback/DemoPlayer.ts
import * as THREE from 'three';
import { DemoFile } from '../types/DemoFile.js';
import { DemoFrame } from '../types/DemoFrame.js';
import { GhostManager } from './GhostManager.js';
import { Interpolator, InterpolationMode } from './Interpolator.js';

/**
 * Playback state.
 */
export enum PlaybackState {
  Stopped,
  Playing,
  Paused,
}

/**
 * Demo player for replaying recorded gameplay.
 * Handles playback controls, interpolation, and ghost object management.
 */
export class DemoPlayer {
  private scene: THREE.Scene;
  private camera: THREE.Camera;
  private demoFile: DemoFile | null = null;
  private ghostManager: GhostManager;
  private playbackState: PlaybackState = PlaybackState.Stopped;
  private playbackTime: number = 0;
  private playbackSpeed: number = 1.0;
  private loopPlayback: boolean = false;
  private interpolationMode: InterpolationMode = InterpolationMode.Linear;
  private gravity: number = -20.0;

  // Ghost objects
  private playerGhostId: string = 'player';
  private freeCamera: THREE.PerspectiveCamera | null = null;
  private useFreeCamera: boolean = false;

  // Events
  private onPlaybackStartedCallbacks: (() => void)[] = [];
  private onPlaybackStoppedCallbacks: (() => void)[] = [];
  private onPlaybackPausedCallbacks: (() => void)[] = [];
  private onPlaybackResumedCallbacks: (() => void)[] = [];
  private onPlaybackTimeChangedCallbacks: ((time: number) => void)[] = [];

  constructor(scene: THREE.Scene, camera: THREE.Camera) {
    this.scene = scene;
    this.camera = camera;
    this.ghostManager = new GhostManager(scene);
  }

  /**
   * Load a demo file for playback.
   */
  loadDemo(demo: DemoFile): void {
    this.stop();
    this.demoFile = demo;
    this.playbackTime = 0;
    console.log(`[DemoPlayer] Loaded demo: ${demo.header.description} (${demo.header.duration}s)`);
  }

  /**
   * Start playback.
   */
  play(): void {
    if (!this.demoFile) {
      console.warn('[DemoPlayer] No demo loaded');
      return;
    }

    if (this.playbackState === PlaybackState.Playing) {
      return;
    }

    this.playbackState = PlaybackState.Playing;

    // Setup replay scene
    this.setupReplayScene();

    // Notify callbacks
    for (const callback of this.onPlaybackStartedCallbacks) {
      callback();
    }

    console.log('[DemoPlayer] Playback started');
  }

  /**
   * Pause playback.
   */
  pause(): void {
    if (this.playbackState !== PlaybackState.Playing) {
      return;
    }

    this.playbackState = PlaybackState.Paused;

    // Notify callbacks
    for (const callback of this.onPlaybackPausedCallbacks) {
      callback();
    }

    console.log('[DemoPlayer] Playback paused');
  }

  /**
   * Resume playback.
   */
  resume(): void {
    if (this.playbackState !== PlaybackState.Paused) {
      return;
    }

    this.playbackState = PlaybackState.Playing;

    // Notify callbacks
    for (const callback of this.onPlaybackResumedCallbacks) {
      callback();
    }

    console.log('[DemoPlayer] Playback resumed');
  }

  /**
   * Stop playback.
   */
  stop(): void {
    if (this.playbackState === PlaybackState.Stopped) {
      return;
    }

    this.playbackState = PlaybackState.Stopped;
    this.playbackTime = 0;

    // Cleanup replay scene
    this.cleanupReplayScene();

    // Notify callbacks
    for (const callback of this.onPlaybackStoppedCallbacks) {
      callback();
    }

    console.log('[DemoPlayer] Playback stopped');
  }

  /**
   * Seek to a specific time.
   */
  seek(time: number): void {
    if (!this.demoFile) {
      return;
    }

    this.playbackTime = Math.max(0, Math.min(time, this.demoFile.header.duration));
    this.updatePlayback();

    // Notify callbacks
    for (const callback of this.onPlaybackTimeChangedCallbacks) {
      callback(this.playbackTime);
    }
  }

  /**
   * Update playback. Call this every frame.
   * @param deltaTime - Time since last frame in seconds
   */
  update(deltaTime: number): void {
    if (this.playbackState !== PlaybackState.Playing || !this.demoFile) {
      return;
    }

    // Advance playback time
    this.playbackTime += deltaTime * this.playbackSpeed;

    // Check for end of demo
    if (this.playbackTime >= this.demoFile.header.duration) {
      if (this.loopPlayback) {
        this.playbackTime = 0;
      } else {
        this.stop();
        return;
      }
    }

    this.updatePlayback();

    // Notify callbacks
    for (const callback of this.onPlaybackTimeChangedCallbacks) {
      callback(this.playbackTime);
    }
  }

  /**
   * Update playback state (apply frames to ghosts).
   */
  private updatePlayback(): void {
    if (!this.demoFile) {
      return;
    }

    // Find frame at current time
    const frame = this.findFrameAtTime(this.playbackTime);

    if (frame) {
      // Update player ghost
      this.ghostManager.updateGhostFromFrame(this.playerGhostId, frame);
    }
  }

  /**
   * Find the frame at a specific time using binary search.
   */
  private findFrameAtTime(time: number): DemoFrame | null {
    if (!this.demoFile || this.demoFile.frames.length === 0) {
      return null;
    }

    const frames = this.demoFile.frames;

    // Binary search
    let left = 0;
    let right = frames.length - 1;

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      const frame = frames[mid];

      if (frame.timestamp === time) {
        return frame;
      } else if (frame.timestamp < time) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    // Return the closest frame
    if (left >= frames.length) {
      return frames[frames.length - 1];
    }
    if (right < 0) {
      return frames[0];
    }

    // Return the frame with the closest timestamp
    const leftFrame = frames[left];
    const rightFrame = frames[right];

    if (Math.abs(leftFrame.timestamp - time) < Math.abs(rightFrame.timestamp - time)) {
      return leftFrame;
    } else {
      return rightFrame;
    }
  }

  /**
   * Setup replay scene (create ghosts, free camera).
   */
  private setupReplayScene(): void {
    if (!this.demoFile) {
      return;
    }

    // Create player ghost
    this.ghostManager.createGhost(this.playerGhostId);

    // Set initial state
    const initialFrame = this.demoFile.frames[0];
    if (initialFrame) {
      this.ghostManager.updateGhostFromFrame(this.playerGhostId, initialFrame);
    }

    // Create free camera (optional)
    if (this.useFreeCamera) {
      this.freeCamera = new THREE.PerspectiveCamera(
        75,
        window.innerWidth / window.innerHeight,
        0.1,
        1000
      );
      this.freeCamera.position.copy(this.demoFile.playerStartPosition);
      this.freeCamera.position.y += 2; // Offset for eye level
    }
  }

  /**
   * Cleanup replay scene (destroy ghosts, free camera).
   */
  private cleanupReplayScene(): void {
    this.ghostManager.removeAllGhosts();
    this.freeCamera = null;
  }

  /**
   * Get the current playback state.
   */
  getPlaybackState(): PlaybackState {
    return this.playbackState;
  }

  /**
   * Get the current playback time.
   */
  getPlaybackTime(): number {
    return this.playbackTime;
  }

  /**
   * Get the demo duration.
   */
  getDuration(): number {
    return this.demoFile?.header.duration ?? 0;
  }

  /**
   * Get the playback progress (0-1).
   */
  getProgress(): number {
    if (!this.demoFile || this.demoFile.header.duration === 0) {
      return 0;
    }
    return this.playbackTime / this.demoFile.header.duration;
  }

  /**
   * Get the ghost manager.
   */
  getGhostManager(): GhostManager {
    return this.ghostManager;
  }

  /**
   * Get the free camera (if enabled).
   */
  getFreeCamera(): THREE.PerspectiveCamera | null {
    return this.freeCamera;
  }

  // Configuration setters
  setPlaybackSpeed(speed: number): void {
    this.playbackSpeed = Math.max(0.1, Math.min(10, speed));
  }

  setLoopPlayback(loop: boolean): void {
    this.loopPlayback = loop;
  }

  setInterpolationMode(mode: InterpolationMode): void {
    this.interpolationMode = mode;
    this.ghostManager.setInterpolationMode(mode);
  }

  setGravity(gravity: number): void {
    this.gravity = gravity;
    this.ghostManager.setGravity(gravity);
  }

  setUseFreeCamera(use: boolean): void {
    this.useFreeCamera = use;
  }

  // Event subscription
  onPlaybackStarted(callback: () => void): void {
    this.onPlaybackStartedCallbacks.push(callback);
  }

  onPlaybackStopped(callback: () => void): void {
    this.onPlaybackStoppedCallbacks.push(callback);
  }

  onPlaybackPaused(callback: () => void): void {
    this.onPlaybackPausedCallbacks.push(callback);
  }

  onPlaybackResumed(callback: () => void): void {
    this.onPlaybackResumedCallbacks.push(callback);
  }

  onPlaybackTimeChanged(callback: (time: number) => void): void {
    this.onPlaybackTimeChangedCallbacks.push(callback);
  }
}
